test/integration/framework/etcd.go

```diff
@@ -136,35 +138,88 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
 	}
 	args = append(args, customFlags...)
 	cmd := exec.CommandContext(ctx, etcdPath, args...)
-	if output == nil {
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-	} else {
-		cmd.Stdout = output
-		cmd.Stderr = output
+
+	// Always filter the etcd output. This allows us to get rid of known harmless messages and (more importantly)
+	// to detect the "listen tcp 127.0.0.1:37803: bind: address already in use" error.
+	reader, err := cmd.StderrPipe()
+	if err != nil {
+		return "", nil, fmt.Errorf("prepared etcd command's stderr pipe: %w", err)
 	}
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		buffer := make([]byte, 100*1024)
+		for {
+			n, err := reader.Read(buffer)
+			// Unfortunately in practice we get an untyped errors.errorString wrapped in an os.Path error,
+			// so we have to fall back to text matching.
+			if errors.Is(err, io.EOF) || err != nil && err.Error() == "read |0: file already closed" {
+				return
+			}
+			if err != nil {
+				logger.Error(err, "read etcd output")
+				return
+			}
+			if n == 0 {
+				continue
+			}
+			dec := json.NewDecoder(bytes.NewBuffer(buffer[0:n]))
+			for {
+				// Try to parse as JSON object. If we get anything that isn't JSON or an object, we just dump the remaining output.
+				var msg map[string]any
+				err := dec.Decode(&msg)
+				if err == io.EOF {
+					// all done
+					break
+				}
+				if err != nil {
+					offset := int(dec.InputOffset())
+					if offset < n {
+						logger.Info("etcd output", "msg", string(buffer[0:n]))
+					}
+					continue
+				}
+
+				// Skip harmless messages.
+				msgStr, _ := msg["msg"].(string)
+				if hiddenEtcdMessages.Has(msgStr) {
+					continue
+				}
+
+				// Log it using structured logging.
+				kvs := make([]any, 0, 2*len(msg))
+				for key, value := range msg {
+					kvs = append(kvs, key, value)
+				}
+				logger.Info("etcd output", kvs...)
+			}
+		}
+	}()
+
 	stop := func() {
 		// try to exit etcd gracefully
 		defer cancel()
 		cmd.Process.Signal(syscall.SIGTERM)
-		var wg sync.WaitGroup
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			select {
 			case <-ctx.Done():
-				klog.Infof("etcd exited gracefully, context cancelled")
+				logger.V(6).Info("etcd exited gracefully, context cancelled")
 			case <-time.After(5 * time.Second):
-				klog.Infof("etcd didn't exit in 5 seconds, killing it")
+				logger.Info("etcd didn't exit in 5 seconds, killing it")
 				cancel()
 			}
 		}()
 		err := cmd.Wait()
+		logger.V(2).Info("etcd exited", "err", err)
+		// Tell goroutine that we are done.
+		cancel()
```

Comment:
Not canceling the context before waiting for the goroutine to exit caused the goroutine to always run into the timeout and the "etcd didn't exit in 5 seconds, killing it" log message.